File: src/main/java/com/commercehub/core/state/annotations/AnnotationBasedClientImplProvider.java
package com.commercehub.core.state.annotations;

import com.commercehub.core.state.StateAction;
import com.commercehub.core.state.Trigger;
import org.reflections.Reflections;

import java.lang.annotation.Annotation;
import java.util.HashMap;
import java.util.Set;

public class AnnotationBasedClientImplProvider implements com.commercehub.core.state.ClientImplProvider {
    private final HashMap<String, Class> triggers;
    private final HashMap<String, Class> actions;

    public AnnotationBasedClientImplProvider(){
        Reflections ref = new Reflections("com");
        Set<Class<?>> types = ref.getTypesAnnotatedWith(RefName.class);

        triggers = new HashMap<String, Class>();
        actions = new HashMap<String, Class>();

        scanForRefNames(types);
    }

    private void scanForRefNames(Set<Class<?>> types) {
        for (Class c: types){
            for (Annotation a: c.getAnnotations()) {
                if (a instanceof RefName) {
                    RefName r = (RefName) a;
                    if (StateAction.class.isAssignableFrom(c)) {
                        actions.put(r.value(), c);
                    }
                    if (Trigger.class.isAssignableFrom(c)) {
                        triggers.put(r.value(), c);
                    }
                }
            }
        }
    }

    @Override
    public Class getTriggerClass(String name){
        return triggers.get(name);
    }

    @Override
    public Class getActionClass(String name){
        return actions.get(name);
    }
}

File: src/main/scala/pulleys/annotations/AnnotationBasedClientImplProvider.scala
package pulleys.annotations

import org.reflections.scanners.SubTypesScanner
import org.reflections.scanners.TypeAnnotationsScanner
import org.reflections.util.ClasspathHelper
import org.reflections.util.ConfigurationBuilder
import org.reflections.util.FilterBuilder
import pulleys.StateAction
import pulleys.Trigger
import org.reflections.Reflections
import pulleys.ClientImplProvider
import java.lang.annotation.Annotation
import java.util
import java.util.concurrent.Executors

class AnnotationBasedClientImplProvider() extends ClientImplProvider {
  val ref: Reflections = new Reflections(new ConfigurationBuilder().setUrls(ClasspathHelper.forJavaClassPath).setExecutorService(Executors.newFixedThreadPool(Runtime.getRuntime.availableProcessors)))
  val types: util.Set[Class[_]] = ref.getTypesAnnotatedWith(classOf[RefName].asInstanceOf[Annotation])

  val triggers = new util.HashMap[String, Class[_]]
  val actions = new util.HashMap[String, Class[_]]

  import scala.collection.JavaConversions._
  for (c <- types) {
    for (a <- c.getAnnotations) {
      if (a.isInstanceOf[RefName]) {
        val r: RefName = a.asInstanceOf[RefName]
        if (classOf[StateAction].isAssignableFrom(c)) {
          actions.put(r.value, c)
        }
        if (classOf[Trigger].isAssignableFrom(c)) {
          triggers.put(r.value, c)
        }
      }
    }
  }

  def getTriggerClass(name: String): Class[_] = triggers.get(name)

  def getActionClass(name: String): Class[_] =  actions.get(name)
}